File: src/main/java/puzzles/year23/day12/Part1.java
package puzzles.year23.day12;

import util.Puzzle;

import java.util.*;
import java.util.regex.Pattern;

public class Part1 implements Puzzle<Integer> {
    @Override
    public Integer getAnswer(String inputString) {
        String[] lines = inputString.split(System.lineSeparator());

        int total = 0;

        for (String line : lines) {
            String[] lineComponents = line.split(" ");
            String pattern = lineComponents[0];
            int[] repeat = Arrays.stream(lineComponents[1].split(",")).mapToInt(Integer::parseInt).toArray();
            Set<String> seen = new HashSet<>();
            Queue<String> queue = new LinkedList<>();
            queue.offer(pattern);
            int possibilitiesCount = 0;

            while (!queue.isEmpty()) {
                String currentLine = queue.poll();
                seen.add(currentLine);

                if (isCorrect(currentLine, repeat)) {
                    possibilitiesCount++;
                    continue;
                }

                String[] possibilities = getPossibilities(currentLine, repeat);

                for (String possibility : possibilities) {
                    if (!seen.contains(possibility)) {
                        queue.offer(possibility);
                    }
                }
            }

            System.out.println(possibilitiesCount);
            total += possibilitiesCount;
        }

        return total;
    }

    private String[] getPossibilities(String line, int[] repeat) {
        if (!isPossible(line, repeat)) {
            return new String[]{};
        }

        List<String> possibilities = new ArrayList<>();

        int index = line.indexOf("?");

        if (index == -1) {
            return new String[]{line};
        }

        possibilities.add(line.replaceFirst("\\?", "."));
        possibilities.add(line.replaceFirst("\\?", "#"));

        return possibilities.toArray(String[]::new);
    }

    private boolean isPossible(String line, int[] repeat) {
        int count = 0;
        int index = line.indexOf("?");

        if (index == -1) return true;

        String[] lineArray = Arrays.stream(line.substring(0, index).split("\\.")).filter(s -> !s.isBlank()).toArray(String[]::new);

        for (int i = 0; i < Math.min(lineArray.length - 1, repeat.length); i++) {
            if (repeat[i] != lineArray[i].length()) {
                return false;
            }
        }

        for (char c : line.toCharArray()) {
            if (c != '.') count++;
        }

        return count >= Arrays.stream(repeat).sum();
    }

    private boolean isCorrect(String line, int[] repeat) {
        String[] lineArray = Arrays.stream(line.split("\\.")).filter(s -> !s.isBlank()).toArray(String[]::new);

        if (lineArray.length != repeat.length) return false;

        for (int i = 0; i < repeat.length; i++) {
            if (lineArray[i].length() != repeat[i]) {
                return false;
            }
        }

        return true;
    }
}

File: src/main/java/puzzles/year25/day4/Part1.kt
package puzzles.year25.day4

import util.Puzzle
import kotlin.CharArray
import kotlin.Int
import kotlin.collections.List

class Part1 : Puzzle<Int?> {
    override fun getAnswer(inputString: String): Int {
        val map = inputString.trim().lines().map{ it.toCharArray()}
        var count = 0

        for (y in map.indices) {
            for (x in map[0].indices) {
                if (map[y][x] == '@' && canPickup(map, x, y)) {
                    count++
                }
            }
        }

        return count
    }

    fun canPickup(map: List<CharArray>, x: Int, y: Int): Boolean {
        var count = 0
        for (i in -1..1) {
            for (j in -1..1) {
                if (i == 0 && j == 0) continue

                if (getValue(map, x + i, y + j) == 1) {
                    if (++count >= 4) return false
                }
            }
        }

        return true
    }

    fun getValue(map: List<CharArray>, x: Int, y: Int): Int {
        if (x < 0 || x >= map[0].size || y < 0 || y >= map.size) {
            return 0
        } else if (map[y][x] == '@') {
            return 1
        }

        return 0
    }
}
